src/app/qgsrelationadddlg.cpp

```diff
@@ -14,79 +14,260 @@
+  // row 5: strength
+  QLabel *strengthLabel = new QLabel( tr( "Relationship strength" ), this );
+  layout->addWidget( strengthLabel, 5, 1 );
+  // col 2
+  mStrengthCombobox = new QComboBox( this );
+  mStrengthCombobox->addItem( "Association", QVariant::fromValue( QgsRelation::RelationStrength::Association ) );
+  mStrengthCombobox->addItem( "Composition", QVariant::fromValue( QgsRelation::RelationStrength::Composition ) );
+  mStrengthCombobox->setToolTip( QStringLiteral( "On composition, the child features will be duplicated too.\n"
+                                 "Duplications are made by the feature duplication action.\n"
+                                 "The default actions are activated in the Action section of the layer properties." ) );
+  layout->addWidget( mStrengthCombobox, 5, 2 );
+
+  // row 6: button box
+  mButtonBox = new QDialogButtonBox( this );
+  mButtonBox->setOrientation( Qt::Horizontal );
+  mButtonBox->setStandardButtons( QDialogButtonBox::Cancel | QDialogButtonBox::Ok );
+  connect( mButtonBox, &QDialogButtonBox::accepted, this, &QgsRelationAddDlg::accept );
+  connect( mButtonBox, &QDialogButtonBox::rejected, this, &QgsRelationAddDlg::reject );
+  layout->addWidget( mButtonBox, 7, 1, 1, 2 );
+
+  // set layout
+  layout->setColumnMinimumWidth( 0, 30 );
+  layout->setColumnStretch( 1, 1 );
+  layout->setColumnStretch( 2, 1 );
+  layout->setRowStretch( 6, 2 );
+  setLayout( layout );
+
+  addFieldPairWidget();
+  addFieldPairWidget();
 
-  mTxtRelationId->setPlaceholderText( tr( "[Generated automatically]" ) );
   checkDefinitionValid();
 
-  connect( mCbxReferencingLayer, &QgsMapLayerComboBox::layerChanged, this, &QgsRelationAddDlg::checkDefinitionValid );
-  connect( mCbxReferencingField, &QgsFieldComboBox::fieldChanged, this, &QgsRelationAddDlg::checkDefinitionValid );
-  connect( mCbxReferencedLayer, &QgsMapLayerComboBox::layerChanged, this, &QgsRelationAddDlg::checkDefinitionValid );
-  connect( mCbxReferencedField, &QgsFieldComboBox::fieldChanged, this, &QgsRelationAddDlg::checkDefinitionValid );
+  connect( mReferencingLayerCombobox, &QgsMapLayerComboBox::layerChanged, this, &QgsRelationAddDlg::checkDefinitionValid );
+  connect( mReferencedLayerCombobox, &QgsMapLayerComboBox::layerChanged, this, &QgsRelationAddDlg::checkDefinitionValid );
+}
+
+void QgsRelationAddDlg::addFieldPairWidget()
+{
+  int index = mFieldPairWidgets.count();
+  QgsFieldPairWidget *w = new QgsFieldPairWidget( index, this );
+  w->setReferencingLayer( mReferencingLayerCombobox->currentLayer() );
+  w->setReferencedLayer( mReferencedLayerCombobox->currentLayer() );
+  mFieldPairsLayout->addWidget( w );
+  mFieldPairWidgets << w;
+
+  connect( w, &QgsFieldPairWidget::configChanged, this, &QgsRelationAddDlg::checkDefinitionValid );
+  connect( w, &QgsFieldPairWidget::pairDisabled, this, [ = ]() {emit fieldPairRemoved( w );} );
```

Comment:
Wouldn't be better to use pairAdded and pairRemoved names instead of enabled/disabled?